crates/storage-query-datafusion/src/remote_query_scanner_client.rs

```diff
@@ -0,0 +1,229 @@
+
+use crate::{decode_record_batch, encode_schema};
+use restate_core::network::rpc_router::RpcRouter;
+use restate_core::network::{Incoming, MessageRouterBuilder, Networking, TransportConnect};
+use restate_core::TaskCenter;
+use restate_types::identifiers::{PartitionId, PartitionKey};
+use restate_types::net::remote_query_scanner::{
+    RemoteQueryScannerClose, RemoteQueryScannerClosed, RemoteQueryScannerNext,
+    RemoteQueryScannerNextResult, RemoteQueryScannerOpen, RemoteQueryScannerOpened,
+};
+use restate_types::NodeId;
+
+// ----- rpc service definition -----
+
+#[async_trait]
+pub trait RemoteScannerService: Send + Sync + Debug + 'static {
+    async fn open(
+        &self,
+        peer: NodeId,
+        req: RemoteQueryScannerOpen,
+    ) -> Result<RemoteQueryScannerOpened, DataFusionError>;
+
+    async fn next_batch(
+        &self,
+        peer: NodeId,
+        req: RemoteQueryScannerNext,
+    ) -> Result<RemoteQueryScannerNextResult, DataFusionError>;
+
+    async fn close(
+        &self,
+        peer: NodeId,
+        req: RemoteQueryScannerClose,
+    ) -> Result<RemoteQueryScannerClosed, DataFusionError>;
+}
+
+// ----- service proxy -----
+pub fn create_remote_scanner_service<T: TransportConnect>(
+    network: Networking<T>,
+    task_center: TaskCenter,
+    router_builder: &mut MessageRouterBuilder,
+) -> Arc<dyn RemoteScannerService> {
+    Arc::new(RemoteScannerServiceProxy::new(
+        network,
+        task_center,
+        router_builder,
+    ))
+}
+
+// ----- datafusion remote scan -----
+
+/// Given an implementation of a remote ScannerService, this function
+/// creates a Datafusion's [[SendableRecordBatchStream]] that transports
+/// record batches via the RemoteScannerService API.
+pub fn remote_scan_as_datafusion_stream(
+    service: Arc<dyn RemoteScannerService>,
+    target_node_id: NodeId,
+    partition_id: PartitionId,
+    range: RangeInclusive<PartitionKey>,
+    table_name: String,
+    projection_schema: SchemaRef,
+) -> SendableRecordBatchStream {
+    let mut builder = RecordBatchReceiverStream::builder(projection_schema.clone(), 2);
+
+    let tx = builder.tx();
+
+    let task = async move {
+        //
+        // get a scanner id
+        //
+        let open_request = RemoteQueryScannerOpen {
+            partition_id,
+            range,
+            table: table_name,
+            projection_schema_bytes: encode_schema(&projection_schema),
+        };
+
+        let RemoteQueryScannerOpened::Success { scanner_id } =
+            service.open(target_node_id, open_request).await?
+        else {
+            Err(DataFusionError::Internal(
+                "Unable to open a remote scanner".to_string(),
+            ))?
+        };
+
+        //
+        // loop while we have record_batch coming in
+        //
+        loop {
+            let req = RemoteQueryScannerNext { scanner_id };
+            let batch = match service.next_batch(target_node_id, req).await? {
```

Comment:
Should we also try to close the scanner if we fail on next_batch with an error? I guess those errors are resulting from network errors, right?